src/clippy-js-proxy.c

```diff
@@ -0,0 +1,675 @@
+                                   jsc_value_to_double (value),
+                                   G_PARAM_READWRITE);
+    }
+  else
+    {
+      return NULL;
+    }
+
+  return pspec;
+}
+
+/**
+ * clippy_js_proxy_properties_from_js:
+ * @webview: A WebKit web view
+ * @object: the object to get properties from
+ *
+ * Get the list of properties from a JavaScript object
+ *
+ * Returns: a new array of GParamSpec structs
+ */
+GParamSpec **
+clippy_js_proxy_properties_from_js (WebKitWebView *webview,
+                                    const gchar   *object)
+{
+  GArray *array = g_array_new (FALSE, FALSE, sizeof (GParamSpec *));
+  WebKitJavascriptResult *result;
+  JSCValue *value;
+
+  if (_js_run_printf (webview, NULL, &result, "%s", object))
+    return NULL;
+
+  if ((value = webkit_javascript_result_get_js_value (result)) &&
+      jsc_value_is_object (value))
+    {
+      g_auto (GStrv) properties = NULL;
+      GParamSpec *pspec = NULL;
+      gint i;
+
+      /* First property is reserved for GObject */
+      g_array_append_val (array, pspec);
+
+      /* Get all JS properties */
+      properties = jsc_value_object_enumerate_properties (value);
+
+      for (i = 0; properties && properties[i]; i++)
+        {
+          const gchar *property = properties[i];
+          g_autoptr (JSCValue) val = jsc_value_object_get_property (value, property);
+
+          pspec = param_spec_new_from_jsc (property, val);
+          g_array_append_val (array, pspec);
+        }
+
+      /* Make it null terminated */
+      pspec = NULL;
+      g_array_append_val (array, pspec);
+    }
+  else
+    {
+      g_warning ("Error running javascript: unexpected return value");
+    }
+
+  webkit_javascript_result_unref (result);
+
+  return (GParamSpec **) g_array_free (array, FALSE);
+}
+
+/* Public API */
+
+/**
+ * clippy_js_proxy_class_new:
+ * @name: the object class name
+ * @properties: Null terminated array of parameter specs
+ *
+ * This function creates a GObject class named @name with properties defined by
+ * @properties param specs array.
+ * Keep in mind the first property is reserved for GObject
+ *
+ * Returns: a new GType
+ */
+GType
+clippy_js_proxy_class_new (const gchar *name, GParamSpec **properties)
+{
+  GTypeInfo *type_info;
+
+  g_return_val_if_fail (name != NULL, 0);
+
+  type_info = g_new0 (GTypeInfo, 1);
+  type_info->class_size = sizeof (ClippyJsProxyDerivedClass);
+  type_info->class_init = (GClassInitFunc) clippy_js_proxy_derived_class_init;
+  type_info->class_data = properties;
+
+  type_info->instance_size = sizeof (ClippyJsProxyDerived);
+  type_info->instance_init = (GInstanceInitFunc) clippy_js_proxy_derived_init;
+
+  return g_type_register_static (CLIPPY_TYPE_JS_PROXY, name, type_info, 0);
+}
+
+static gchar *
+get_class_name_from_object (const gchar *object)
+{
+  g_auto (GStrv) tokens = g_strsplit (object, "_", 0);
+  GString *retval = g_string_new ("ClippyJSProxy");
+  gint i;
+
+  for (i = 0; tokens[i]; i++)
```

Comment:
Nitpick: Declare the `gint i` as part of the for.

Reply:
I also think a comment saying this turns the object name from snake_case into CamelCase and appends it to `ClippyJsProxy` would be nice.